File: src/main/java/org/springframework/cloud/config/monitor/PropertyPathEndpoint.java
package org.springframework.cloud.config.monitor;

import org.springframework.cloud.bus.event.Destination;
import org.springframework.cloud.bus.event.RefreshRemoteApplicationEvent;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.context.ApplicationEventPublisherAware;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.util.StringUtils;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

@RestController
@RequestMapping(path = "${spring.cloud.config.monitor.endpoint.path:}/monitor")
public class PropertyPathEndpoint implements ApplicationEventPublisherAware {

    public static final String PATH = "path";
    public static final String APPLICATION = "application";
    public static final String EMPTY = "";
    public static final String DASH = "-";
    public static final String SEMI_CLONE = ":";
    public static final String WILD_CARD = "*";
    private ApplicationEventPublisher applicationEventPublisher;

    private final PropertyPathNotificationExtractor extractor;
    private final String busId;

    public PropertyPathEndpoint(PropertyPathNotificationExtractor extractor, String busId) {
        this.extractor = extractor;
        this.busId = busId;
    }

    @Override
    public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
        this.applicationEventPublisher = applicationEventPublisher;
    }

    @PostMapping
    @SuppressWarnings("deprecation")
    public Set<String> notifyByPath(@RequestHeader HttpHeaders headers, @RequestBody Map<String, Object> request) {
        final PropertyPathNotification notification = this.extractor.extract(headers, request);

        if (Objects.isNull(notification)) {
            return Collections.emptySet();
        }


        final Set<String> services = new LinkedHashSet<>();

        for (String path : notification.getPaths()) {
            services.addAll(guessServiceName(path));
        }

        if (Objects.isNull(this.applicationEventPublisher)) {
            return Collections.emptySet();
        }

        for (String service : services) {
            this.applicationEventPublisher
                    .publishEvent(new RefreshRemoteApplicationEvent(this, this.busId, service));
        }

        return services;
    }

    @PostMapping(consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
    public Set<String> notifyByForm(@RequestHeader HttpHeaders headers, @RequestParam(PATH) List<String> request) {
        final Map<String, Object> map = new HashMap<>();
        map.put(PATH, request);
        return notifyByPath(headers, map);
    }

    private Set<String> guessServiceName(String path) {
        final Set<String> services = new LinkedHashSet<>();

        if (Objects.isNull(path)) {
            return services;
        }

        final String stem = Optional.of(path)
                .map(StringUtils::cleanPath)
                .map(StringUtils::getFilename)
                .map(StringUtils::stripFilenameExtension)
                .orElse(EMPTY);

        final AtomicInteger index = new AtomicInteger(stem.indexOf(DASH));

        while (index.get() >= 0) {
            String name = stem.substring(0, index.get());
            String profile = stem.substring(index.get() + 1);

            services.addAll(getServices(name, profile));

            index.set(stem.indexOf(DASH, index.get() + 1));
        }

        if (APPLICATION.equals(stem)) {
            services.add(WILD_CARD);
            return services;
        }

        if (!stem.startsWith(APPLICATION)) {
            services.add(stem);
            return services;
        }

        return services;
    }

    private Set<String> getServices(String name, String profile) {
        final Set<String> services = new HashSet<>();

        if (APPLICATION.equals(name)) {
            services.add(WILD_CARD + SEMI_CLONE + profile);
            return services;
        }

        if (!name.startsWith(APPLICATION)) {
            services.add(name + SEMI_CLONE + profile);
            services.add(name);
            return services;
        }

        return services;
    }

}

File: src/main/java/org/springframework/cloud/config/monitor/PropertyPathEndpoint.kt
package org.springframework.cloud.config.monitor

import java.util.concurrent.atomic.AtomicInteger
import org.springframework.cloud.bus.event.RefreshRemoteApplicationEvent
import org.springframework.context.ApplicationEventPublisher
import org.springframework.context.ApplicationEventPublisherAware
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.util.StringUtils
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping(path = ["\${spring.cloud.config.monitor.endpoint.path:}/monitor"])
class PropertyPathEndpoint(
    private val extractor: PropertyPathNotificationExtractor,
    private val busId: String,
) : ApplicationEventPublisherAware {

    private lateinit var applicationEventPublisher: ApplicationEventPublisher

    override fun setApplicationEventPublisher(applicationEventPublisher: ApplicationEventPublisher) {
        this.applicationEventPublisher = applicationEventPublisher
    }

    @PostMapping(consumes = [MediaType.APPLICATION_FORM_URLENCODED_VALUE])
    fun notifyByForm(@RequestHeader headers: HttpHeaders, @RequestParam(PATH) request: List<String>): Set<String> {
        val map: MutableMap<String, Any> = HashMap()
        map[PATH] = request
        return notifyByPath(headers, map)
    }

    @PostMapping
    @Suppress("DEPRECATION")
    fun notifyByPath(@RequestHeader headers: HttpHeaders, @RequestBody request: Map<String, Any>): Set<String> {
        val notification = extractor.extract(headers, request)

        return notification?.run {
            val services = this.paths
                .flatMap { guessServiceName(it) }
                .toSet()

            services.forEach {
                applicationEventPublisher.publishEvent(RefreshRemoteApplicationEvent(this, busId, it))
            }

            services
        } ?: emptySet()
    }

    private fun guessServiceName(path: String?): Set<String> {
        val services: MutableSet<String> = LinkedHashSet()

        if (path.isNullOrEmpty()) {
            return services
        }

        val stem = path.run {
            val cleanPath = StringUtils.cleanPath(this)
            val filename = StringUtils.getFilename(cleanPath)
            filename?.let { StringUtils.stripFilenameExtension(it) }
        } ?: EMPTY

        val index = AtomicInteger(stem.indexOf(DASH))

        while (index.get() >= 0) {
            val name = stem.substring(0, index.get())
            val profile = stem.substring(index.get() + 1)
            services.addAll(getServices(name, profile))
            index.set(stem.indexOf(DASH, index.get() + 1))
        }

        if (APPLICATION == stem) {
            services.add(WILD_CARD)
            return services
        }

        if (stem.startsWith(APPLICATION)) {
            return services
        }

        services.add(stem)
        return services
    }

    private fun getServices(name: String, profile: String): Set<String> {
        val services: MutableSet<String> = HashSet()

        if (APPLICATION == name) {
            services.add(WILD_CARD + SEMI_CLONE + profile)
            return services
        }

        if (name.startsWith(APPLICATION)) {
            return services
        }

        services.add(name + SEMI_CLONE + profile)
        services.add(name)
        return services
    }

    companion object {
        private const val PATH = "path"
        private const val APPLICATION = "application"
        private const val EMPTY = ""
        private const val DASH = "-"
        private const val SEMI_CLONE = ":"
        private const val WILD_CARD = "*"
    }

}